### ISSUE TITLE
Enable Helm Testing with Values in Karpenter CI

### ISSUE THREAD

**Issue by jonathan-innis** [aws/karpenter#4989]
### Description

**What problem are you trying to solve?**

Right now, we do not have visibility into testing different values that are passed into the helm chart for installation. We are going off of manual testing to validate that a given configuration of values that we support works. We test an installation of the default helm values in our E2E testing but this misses a number of different values combinations that may fail with the pushed helm chart. Most recently, we [had a bug](#4979) with installing the helm chart with `--set webhook.enabled=false`. 

We should consider looking at `helm test` or some other mechanism to test our helm values within CI so that we can be sure that new values combinations that we introduce into the helm chart work as expected and do not fail on install.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jonathan-innis** [comment]
Can look into https://helm.sh/docs/topics/chart_tests/

**Comment by karthikmurali60** [comment]
@jonathan-innis I am interested in working on this issue. How do i get started ?

**Comment by jonathan-innis** [comment]
Hey @karthikmurali60. I'm not personally as familiar with the best practices around testing helm charts. I know that `helm test` works with integration testing for helm, but I'd rather us go the unit testing route _first_ before we start looking at deploying the chart with values to an actual cluster.

I know that @stevehipwell is a bit of an expert here and has a lot of experience with building out helm charts and testing them. He might be a good POC for asking how we should proceed here.

**Comment by stevehipwell** [comment]
@jonathan-innis my recommendation would be to tackle this in a number of stages starting with the simplest and adding complexity as required.

1. Create a values file which enables most things and run it through [kubeconform](https://github.com/yannh/kubeconform) for each supported K8s version - **Tests the validity of the output YAML for most parts of the template that aren't mutually exclusive**
2. Create a values file for each Karpenter configuration and run these through [kubeconform](https://github.com/yannh/kubeconform) for each supported K8s version - **Should test the validity of the output YAML for all parts of the templates**
  a. Add Hem testing hooks to validate the Karpenter install
4. Run chart testing for each Karpenter configuration values file - **Tests the correctness of the KRM objects and Karpenter args**
  a. Add Hem testing hooks to validate the Karpenter install
5. Add a values schema - **Blocks the use on unexpected inputs which could break templating**
6. Add unit tests - **Belt and braces**

Running through _kubeconform_ in this context means passing the output from `helm template` into it. Running multiple values files through chart testing for something like Karpenter would require a custom workflow that ran each test in sequence rather than the default behaviour.

I guess there are other questions here about how Karpenter can report back if it's correctly configured or if we need a fake cloud to test against.

---

For bonus points you could fuzz the chart values and run them through _kubeconform_, but you'd probably want a schema before this.

**Comment by jonathan-innis** [comment]
> Add a values schema

For this one, I agree! I'm going to open a separate issue for this though because this seems like it could be segregated out.

**Comment by stevehipwell** [comment]
> Yeah, I'm less inclined to want to go this deep on checking everything works with the cloudprovider. I'm mainly concerned with ensuring that different configurations of the values files produce valid outputs and that the general install works as a starting point

In this case I think multiple values files run through `kubeconform` would be the best first step, maybe with an additional validation (`yq` based if unit tests add too much complexity) to check the args and anything else that.

**Comment by jonathan-innis** [comment]
@stevehipwell Do you have a pointer to an example where you are doing something like this already that we could use to map the same testing style over here?
